### ISSUE TITLE
Allow commenting on non-pull-request code

### ISSUE THREAD

**Issue by dmugtasimov** [isaacs/github#284]
When reading someone's code I need to ask an author why he/she code that way, not another or ask for clarfication of piece of code for better understanding or suggest an alternative solution for consideration without making pull request. There fore it would useful communication in the context of the code line, so please, allow commenting on any code in repository.

**Comment by pb-cdunn** [comment]
I'm fine with putting all my comments on the latest commit, but I cannot view the entire file. If I click "View", then I get the whole file at that comment, but with no way to comment. All I can do is to use the commit view (with multiple files), and repeatedly repeatedly repeatedly click the little button to expand the code. For a large file, it takes forever!

An "expand all" button would suffice.

**Comment by efgerrol** [comment]
I just ran into that issue today.

**Comment by maiamcc** [comment]
+1 -- would love a way to comment _in a PR_ on code not changed in the commits associated with that PR. (e.g. sometimes new code has rendered old docstrings obsolete and I want to mention this in a comment.)

**Comment by Thorgard** [comment]
Just stumbled across this issue today. I was looking for a similar feature.

**Comment by holtzermann17** [comment]
Here's an example using Hypothesis to "inject" comments.  I realise this isn't what people were asking for but it might work in a pinch.
https://via.hypothes.is/https://gist.github.com/caugner/589cac8dbb07b26b34ff

**Comment by aubergine10** [comment]
It's a PITA to have to track down a commit relevant to the code (especially if it's from many months ago). Would be nice to be able to comment on a line and have github work out which commit it relates to automagically.

**Comment by alexrj** [comment]
I think what we really want is the ability to start a review attached to an entire file at a given commit, and have it listed/linked among issues.

**Comment by pio-davidt** [comment]
+1 -- Someone creates a PR to show me how they have changed the code to achieve a particular purpose. I think there is some code, unchanged in their PR, that should be changed to achieve the purpose of this PR. I would like to be able to comment on this, and for the PR author, the most useful place for me to leave the comment is next to the (as-yet-unchanged) code. In my mental model, the comment relates to this PR, not to an old commit that made the code how it is.

**Comment by bordumb** [comment]
+1 This would be really nice to have. 

Currently, the only way to add line-by-line comments is to go through a formal pull request workflow. Sometimes you want to allow people to more freely push code (ad-hoc/unimportant work). It doesn't make sense to always force people into making pull requests just so to get comments on code.

Any update on this?

**Comment by davidstanke** [comment]
I think of this in the context of an _audit_. Say I have a repo that I've been working on for a while: lots of commits, lots of PRs (or maybe even no PRs, if I'm the sole author). I want to ask a colleague to take a look and give general feedback. Some of their notes may be about project-level design choices, others may be file-level thoughts, and still others may be line-by-line nits. It doesn't matter how the code got to the state it's in; I'm asking for review of current state... the _entire_ current state.

I'm not really sure how to do this. (Other than to create a branch, delete all contents of that branch, then make a PR from the actual repo contents against that blank branch. Which seems rather awkward.)

**Comment by farrukhnajmi** [comment]
The way GitLab does this is very functional and seammless because it does not require hunting down the commit where code you want to comment was changed. Please consider support similar:

https://docs.gitlab.com/ee/user/project/merge_requests/work_in_progress_merge_requests.html

**Comment by vadzim-revolist** [comment]
+1
I need to make comments to the code that I think was _intended_ to be changed, but wasn't.
So I need to comment any line of the file, not only near the changes.

**Comment by aheninger** [comment]
+1
Trying to review a PR that is cleaning up a class of problem in a file, and it missed one occurrence. There's no way to point out the problem where it occurs.

**Comment by eallenOP** [comment]
Marking student work, I want to be able to leave comments all through any given file on the commit closest to the due time (not just on the lines that changed, because often it's just a merge and nothing changed). I guess I could make an issue for each comment that references the line, but that means a lot of back and forth.

**Comment by JStickler** [comment]
Same problem with documentation PRs, sometimes changes in the code require updates in the docs.  It would be nice to be able to point out, in context, where someone missed something in a documentation update.

**Comment by krulik** [comment]
What @davidstanke said.
This post explains how to do it manually:
https://astrofrog.github.io/blog/2013/04/10/how-to-conduct-a-full-code-review-on-github/

**Comment by arturpat** [comment]
Just saw encountered this issue. Anybody found a way to do so? I did what @davidstanke said (delete and re-add) but this is a dirty workaround and it includes using features for totally different purposes than what they were designed for.

**Comment by barbogast** [comment]
![image](https://user-images.githubusercontent.com/941889/64185427-6a331300-ce6d-11e9-8e38-bec1affd6ec7.png)

😢

**Comment by molysgaard** [comment]
This is an obvious issue in my company when doing a code-review of a pull-request.

The diff in a PR might require changes to lines that _are not changed_ in the PR. Since these lines are greyed out in a PR-review it is not possible to comment on them.

I guess the underlying problem is that a PR-review is modeled as a review of a diff, instead of a review of the repository state at a certain commit. This is a fundamental flaw in GitHubs design of a review since you can not guarantee that changes in a diff only affect the lines _in the diff itself_.

A trivial example is renaming a function, but forgetting to rename all usages of it. Why should it not be possible to comment on the call site with the old function name, where the problem actually is located?

**Comment by arturpat** [comment]
@molysgaard Exactly this! It happens often that the change is affecting the lines that are not part of this commit. If it weren't for project requirements I would be on Gerrit without even thinking twice.

**Comment by tskopek** [comment]
We just migrated to Github from Bitbucket. I got so used to doing this in Bitbucket, I never considered this would not be possible in Github.

This is an extremely useful feature, for many of the reasons outlined above.

**Comment by arturpat** [comment]
@tskopek wait is it possible in Bitbucket? I have this issue in Bitbucket. Were you able to work around it?

**Comment by tskopek** [comment]
Hey @arturpat, well, not exactly in the complete way that this Issue requests (where you can comment on any file in the repo), but in Bitbucket you can at least comment on other parts of the files that are part of the PR, outside of the "diff section". In github I am not even able to do this!

Hopefully this helps explain:

![image](https://user-images.githubusercontent.com/56603278/67897757-6baf4d80-fb35-11e9-997b-0534bb6d6f0a.png)

**Comment by arturpat** [comment]
Oh I see what you meant. True, I am able to add comments exactly +/- 10 lines of what was changed. Lines that are further that 10 lines from the changed ones are grayed out and I can't add comments there. Maybe this is something that can be configured in bitbucket config which I don't have access to. Anyway, thank you @tskopek for taking your time to explain :)

**Comment by TechMaz** [comment]
+1 this, would love the ability to comment on other lines in the code in a pull request

**Comment by Exr0n** [comment]
+1 this, are there any updates? Is there an official feature request?

**Comment by neverdieTRX** [comment]
+1 this would be a really useful feature

**Comment by fbidu** [comment]
Just ran into this! Passing by to give a +1

**Comment by Markkop** [comment]
This would help a lot on external Code Reviews, please consider this feature :rocket:

**Comment by markhealy88** [comment]
Azure DevOps allows commenting on any line of code in a changed file within the context of a pull request.  This is a very useful feature, such as highlighting areas that might have been overlooked as part of the aim of the PR

**Comment by aleksey-rezvov** [comment]
I think I solved a similar task:
https://github.com/oktend/system-review-example

**Comment by BilalIshtiaque** [comment]
This would be very helpful, as others mentioned, a lot of time certain code outside the PR context, esp within the same file, need to be updated as per the PR changes, but being unable to leave comments make it very hard to communicate properly

**Comment by tanujkhattar** [comment]
Much needed feature! +1

**Comment by chrisjsewell** [comment]
+1 to bump this up the priority list 😬

**Comment by calculuschild** [comment]
Another vote for this here. I'm surprised this has been here for more than 5 years already.

**Comment by denalimarsh** [comment]
+1 bump, please implement this feature.

**Comment by dubois** [comment]
The current behavior is baked into the design of the github API. When adding a comment, the line is specified relative to the diff being viewed; see [this picture](https://i.stack.imgur.com/t37he.png) from [this SO question](https://stackoverflow.com/questions/41662127).

It's too bad, but I think we're stuck with it.

**Comment by ronakg** [comment]
This is such a basic use-case that comes up so often in our reviews. Someone made a change that applies to multiple places, but they missed a couple. Now there's no way to comment on those 2 places reminding them that they missed.

**Comment by avanathan** [comment]
Even generic comments (for the entire file) is good enough too, like we have on GitHub issues. Feel like this is basic use case & people use some hacky solutions to get around it - where this could be a great feature (for the effort it needs (?)).

**Comment by krisdale** [comment]
+1 - a much-needed feature!

**Comment by geiersbach** [comment]
+1 - a basic and much-needed feature!

**Comment by dgw** [comment]
Reviewed a PR today that needs a new Python requirements entry to work, but can't add a suggestion to the relevant file because the PR author didn't touch it.

Not the sexiest example, but suggestions really are more useful if we can actually _make_ them. 😁

**Comment by muescha** [comment]
yes - sometimes it is more easy to make suggestions than explain which file the PR author has to touch

**Comment by nitheeshas** [comment]
Just got hit with the same limitation. A bit surprised that this use case isn't covered already in GitHub.

**Comment by nitheeshas** [comment]
Just got hit with this limitation. A bit surprised that this use case isn't covered already by GitHub.

**Comment by KjellKod** [comment]
+1 :

**Comment by xuto2** [comment]
+100

**Comment by jd-lacey** [comment]
Ugh just ran into this issue. This is ridiculous and seems like such an arbitrary restriction. It's been almost 6 years. Come on Github, fix this.

**Comment by muescha** [comment]
@jd-lacey i think GitHub is not reading this here ... you should send them a feedback

**Comment by 0xba1a** [comment]
+1

**Comment by kpeluso** [comment]
+1

**Comment by WilliamLQin** [comment]
+1

**Comment by oxygenecore** [comment]
Would be nice to hear official comment on this :thinking:

**Comment by jeffabailey** [comment]
> Would be nice to hear official comment on this 🤔

+1

**Comment by Exr0n** [comment]
It looks like they've finally added this feature?

If you go to any commit, you can scroll to the bottom to add a comment:
![image](https://user-images.githubusercontent.com/25516241/89433310-06594000-d6f7-11ea-8bb2-d66110957d07.png)
or comment on lines of code directly:
![image](https://user-images.githubusercontent.com/25516241/89433366-183ae300-d6f7-11ea-9bb7-963749ef170d.png)

This is what I was looking for, at least. Glad they finally added it :)

**Comment by dgw** [comment]
@Exr0n That feature's existed for _years_. Commenting there is completely disconnected from any PR review.

**Comment by Exr0n** [comment]
Hm, I guess I never noticed it then. @dgw what feature did you have in mind? I thought this matched the description of the original comment for this issue: "allow commenting on any code in repository".

**Comment by dgw** [comment]
@Exr0n Most of us (I think) are looking for a solution to situations like the one I described [here](https://github.com/isaacs/github/issues/284#issuecomment-627617941). That is, when we're reviewing PRs and need to add line notes or suggestions outside the context of the current patch.

**Comment by muescha** [comment]
> If you go to any commit, you can scroll to the bottom to add a comment:

i have done some of this kind of comments in commits, but never get an answer - i am not sure if they really delivered to the code owner

**Comment by oxygenecore** [comment]
Ok, I've finally read the description to this repo. It's not anything official. Though, I could not find any official public bug repository for GitHub cloud. Does anyone know where were should post to have official answer?

**Comment by HighCommander4** [comment]
> Though, I could not find any official public bug repository for GitHub cloud.

I'm not aware of one. (If only they had a convenient platform where they could host their own bug tracker...)

**Comment by muescha** [comment]
@oxygenecore 

- you can contact GitHub direct: https://support.github.com/contact - you get always an answer 👍 
- or use the community, where also official GitHub employees are active: https://github.community/

**Comment by jasonkhadka-kd** [comment]
Still no update on this from GitHub. how can we tag them here?

**Comment by muescha** [comment]
@jasonkhadka-kd this is not a official github repo

**Comment by chrisjsewell** [comment]
>     * you can contact GitHub direct: [support.github.com/contact](https://support.github.com/contact) - you get always an answer 👍
> 
>     * or use the community, where also official GitHub employees are active: [github.community](https://github.community/)

If someone does this, can they post the reply/link here thanks.
It would be nice to know that someone in GitHub is aware of this issue and the amount of upvotes it has here.

**Comment by calculuschild** [comment]
I have contacted them multiple times pointing to this repo. Never got a response. Maybe someone else will have better luck?

**Comment by jason-koppel-avlino** [comment]
+1

**Comment by srozdilsky** [comment]
+1
IMO, this is not just a feature request but a bug (missing basic functionality).  I have always been frustrated by the lack of ability to do this in a github PR.  I have never had an issue in other Peer Review platforms.

**Comment by dabrahams** [comment]
For what it's worth, there's a project called [ReviewNB](https://www.reviewnb.com/) that supports this feature for Jupyter Notebooks in GitHub.  If someone could figure out how to make that work for all files, it would solve this problem.

**Comment by bw2** [comment]
+1

**Comment by hoangnm-1484** [comment]
> aaa

**Comment by lupinitylabs** [comment]
+1. I am surprised that this doesn't work in Github. Especially in a refactoring PR, there is no accurate way of pointing out to someone that they missed something.

**Comment by okwme** [comment]
+1

**Comment by Androz2091** [comment]
I agree this would be really useful, and we should be able to make suggestions too. Someone sent me a PR, it was a complete rewrite of my NPM package in TypeScript and sometimes he forgot to add types and I'm not able to comment on these lines...

**Comment by ToddBradley** [comment]
It's now been 6 years and 3 months. Welcome to 2021, dear readers. As far as I can tell, there has been no progress in GitHub in this area.

**Comment by meliodev** [comment]
`
import  
import  
`

**Comment by chadm-sq** [comment]
Not that you need more justification, but: In Go and some other languages, it is an error to import a module and not use it. A PR that removes the final use of a module, will cause problems up outside of the lines changed in the PR. A good commented PR has to be able to refer to those.

**Comment by bonafidehan** [comment]
Wow surprising you can't add comments on unchanged code. Logic changes in a single PR isn't encapsulated to just changed lines. I'm guessing nobody from GitHub is paying attention to this issue.

**Comment by bhrutledge** [comment]
I found this related post on the GitHub support forum: https://github.community/t/feature-request-review-comment-on-whole-file/135108. If folks know of other ones, could you add them here for reference?

**Comment by bhrutledge** [comment]
There's a Visual Studio Code extension that would facilitate this process: https://marketplace.visualstudio.com/items?itemName=vsls-contrib.codetour. There has been some discussion about adding it to the GitHub UI: https://github.com/microsoft/codetour/issues/10. The latest thinking is that [GitHub CodeSpaces](https://github.com/features/codespaces) will be "the natural mechanism for enabling developers to replay a tour from GitHub." Someone has also created browser extensions for Chrome and Firefox to play tours in GitHub: https://github.com/doctolib/code-tours-github.

**Comment by shadiramadan** [comment]
How is this still an open issue 7 years later. @clarkbw

**Comment by shadiramadan** [comment]
I'm actually starting to regret my orgs switch to github.

**Comment by SylvainTakerkart** [comment]
+1 on the general feature request described in this issue!

**Comment by chadm-sq** [comment]
Unless you're fixing it or adding important non-personal details, please only just vote with the 👍🏻  and such at the top. Commentary noise is a good reason for the owner to close this bug report without addressing it. Small, discrete bug reports, without drama or entitled whining are best.

**Comment by tribbloid** [comment]
+1 and particularly important when combining with https://github.com/marketplace/actions/publish-unit-test-results

**Comment by davidvincze** [comment]
+1

**Comment by ililic** [comment]
How has this been open since 2014 lol 💩

**Comment by chrisstephenmiller** [comment]
👍

**Comment by Levi-Lesches** [comment]
This isn't the official GitHub repo, https://github.com/github/feedback is. Open a discussion [in the general feedback category](https://github.com/github/feedback/discussions/categories/general-feedback) (and link back to here) to get a response from GitHub staff.

**Comment by ting-lan-wang** [comment]
+1

**Comment by chadm-sq** [comment]
This is a private bug. No one from Github watching this or is using this to schedule work.

There are "+1" and smiley buttons, if you want to show support to things on Github, but adding comments like the previous sends email to a hundred people who don't care that much that you agree. If you want to indicate your support in a polite way, find the ":)" at the top.

Everyone else, if these notifications bother you too, please change your "participating" status to "ignored" at the top. https://github.com/isaacs/github/issues/284 I changed mine to avoid getting messages appended to this, and you can too.
